### PR TITLE
Adding timing implementation including interrupts, updated unit tests

### PR DESCRIPTION
**Pull request by flatherz**
The biggest difference is the change I made so that the running an instruction returns a non-zero value representing the number of cycles. The commit history is a mess and includes several code implementations we have out for review. As we get those reviewed and merged I'll update this branch. so the diff looks less gnarly.

### CODE REVIEW

File: emulator.c [hunk 1]
```diff
@@ -380,7 +427,23 @@ execute_instruction(i8080 *cpu, uint8_t opcode)
...
+          {
+            uint16_t address = cpu_read_mem(cpu, cpu->pc + 2); // NOLINT
+            address = address << 8;                            // NOLINT
+            address += cpu_read_mem(cpu, cpu->pc + 1);
+            cpu->pc = address;
+            // NOLINTNEXTLINE
+            return 10; // return num cycles
+          }
+        cpu->pc += 2;    // NOLINT
+        num_cycles = 10; // NOLINT
```
Comment (cliff-bielinski) [inline emulator.c:445]: Interesting how this one has the same cycle count regardless of jump or not compared to 0xc8 which is different. As far as I can tell this is correct.

File: emulator.c [hunk 2]
```diff
@@ -694,8 +777,9 @@ void
 print_state(i8080 *cpu)
 {
   printf("REGISTERS a: 0x%02x b: 0x%02x c: 0x%02x d: 0x%02x e: 0x%02x h: "
-         "0x%02x l: 0x%02x ",
-         cpu->a, cpu->b, cpu->c, cpu->d, cpu->e, cpu->h, cpu->l);
+         "0x%02x l: 0x%02x sp: 0x%02x pc: 0x%02x ",
```
Comment (cliff-bielinski) [inline emulator.c]: Shouldn't this be 0x%04x since sp and pc are 16-bit addresses?

File: shell.c [hunk 3]
```diff
@@ -1,17 +1,25 @@
...
 #include <unistd.h>
 
+#define CLOCK_SPEED_MS 2000
+#define TICK (1000 * (1.0 / 60.0))
+#define CYCLES_PER_TICK (CLOCK_SPEED_MS * TICK)
+
+int run_cpu(i8080 *cpu, int cycles);
+int pflag = 0;
+int dflag = 0;
+int opt;
```
Comment (cliff-bielinski) [inline shell.c]: Does opt need to be global? I believe it's only ever used in main while calling getopt.

File: shell.c [hunk 4]
```diff
@@ -60,45 +68,86 @@ main(int argc, char *argv[])
...
-      // fetch_decode_execute(&cpu)
+      if ((SDL_GetTicks() - last_tick) > TICK) // NOLINT
+        {
+          if (pflag)
+            {
+              printf("Current Tick: %d\n", SDL_GetTicks());
+            }
+
+          // run first half of tick cycles
+          cycle_offset = run_cpu(&cpu, num_cycles - abs(cycle_offset));
```
Comment (cliff-bielinski) [inline shell.c:88]: cycle_offset should always be 0 or a negative number, correct?

File: tests.c [hunk 5]
```diff
@@ -450,6 +450,65 @@ test_opcode_0xc6(void)
...
+  cpu_write_mem(&cpu, cpu.pc + 1, 0x3a);        // NOLINT
+  cpu_write_mem(&cpu, cpu.pc + 2, 0xa9);        // NOLINT
+  update_zero_flag(&cpu, 0);                    // no remainder
+  code_found = execute_instruction(&cpu, 0xca); // NOLINT
+  CU_ASSERT(code_found >= 0);
+  CU_ASSERT(cpu.pc == 0xa93a); // NOLINT
+
+  // cleanup
+  cpu_write_mem(&cpu, 0x0001, 0x00); // NOLINT
+  cpu_write_mem(&cpu, 0x0002, 0x00); // NOLINT
```
Comment (cliff-bielinski) [inline tests.c]: Cleaning up wrong memory - I think this is fixed in the other PR?

**Review by cliff-bielinski — CHANGES_REQUESTED** [review]
Really nice job overall. A few comments. I think only one thing that needs to be changed (print_state for pc and sp).

File: emulator.c (hunk 2 above)
Comment (flatherz) [reply emulator.c]: Yep! I'll push that change soon.

File: shell.c (hunk 3 above)
Comment (flatherz) [reply shell.c]: I think that's correct. I'll move it back.

File: shell.c (hunk 4 above)
Comment (flatherz) [reply shell.c:88]: Yes, I used abs() purely for the sake of readability. I had num_cycles + cycle_offset and realized I would confuse myself probably 20 minutes later :^]

File: tests.c (hunk 5 above)
Comment (flatherz) [reply tests.c]: That's correct.

File: emulator.c (hunk 1 above)
Comment (flatherz) [reply emulator.c:445]: Yeah our sources definitely seem to indicate when this varies...

**Review by cliff-bielinski — APPROVED** [review]

**Review by Roamingturtl — APPROVED** [review]

**Pull request merged**